### ISSUE TITLE
Machine Learning

### ISSUE THREAD

**Issue by jordan-wirth** [redoverture/PACE#6]
Ideally we should train a SciKitLearn model to predict area congestion for the final project. The model should take in a day of week and time of day and predict a level of congestion (which still needs to be defined better than the simple set maximum).

**Comment by jordan-wirth** [comment]
Good news! I've got a proposed model to learn our patterns. Check out 3859fd2b62bdc556475bea811feb3a8654b2ae45 for the code. I found that SVC RBF supervised learning was able to fit our data pretty well. It has three parameters which I tuned by hand to get around 85% accuracy on just 3 days of data - `C=1, gamma=0.1, epsilon=.01`.

To make our predictions, I removed the date from the data and just had the minute number, absolute from 0-1440 through a day. So each day contributes another point to each minute (every 10 minutes, that is). That looks like this:
# Raw Data, Pre-Processed
![image](https://user-images.githubusercontent.com/20449792/56068716-ce81f880-5d4d-11e9-935d-7c6dd7bf5c5d.png)

Next, I trained our model on the data. I then asked it to generate a congestion prediction for *every minute* (even though we gave it every 10, as we collected it. That looks like this:
# Predicted Data from SVC RBF
![image](https://user-images.githubusercontent.com/20449792/56068768-08eb9580-5d4e-11e9-8617-c081f0f0c601.png)
![image](https://user-images.githubusercontent.com/20449792/56068774-0ee17680-5d4e-11e9-9ad1-711f5d60a2c6.png)

Personally, I think this looks pretty good. Once we have a larger data set, I'd like to cross-validate the training and see if it gets better!

**Comment by jordan-wirth** [comment]
Even better news! The way I did things before was weird and apparently over-fitting a lot. So I tackled the problem anew today and got something much more reasonable.

![image](https://user-images.githubusercontent.com/20449792/56100618-11c49e80-5ee9-11e9-8d6c-f051217b6cd0.png)

Here's the code to produce this:
`import pace.analysis.read as rd`
`import pace.analysis.learn as lrn`
`import pace.analysis.plot as pplt`
`telemetry = rd.pickle_jar(rd.find_all_pickles('data/'))`
`train = lrn.preprocess(telemetry)`
`k = lrn.model_svr(*train)`
`pplt.eval_model(*train, k).show()`

This model is 84.35% accurate in training, and 85.10% accurate in testing. I realized after fighting with ~60% accuracy for awhile that *weekends* were affecting the accuracy most! It makes sense - people behave differently on the weekends than they do during classes. This will be a key point in our presentation. This tool can be *much* more accurate if it collects data for a large series of 'Mondays', etc. We just didn't have the time to collect that much data. So, I trashed the weekend data and the set became much more predictable!

I also tried a Polynomial characteristic pre-processing, with Ridge linear regression. It worked, but at around 75% accuracy, so not really that useful. Here's what it looks like:
![model_poly_decision](https://user-images.githubusercontent.com/20449792/56100670-a4653d80-5ee9-11e9-9a1b-7b425466ca1f.png)

I think this is presentable as it stands. The only other improvement we could make is fine grained analysis of parameters, probably using GridSearch. I don't have time to do this as of now, so if anyone would like to tackle this feel free! For now, however, since the results are acceptable I'll close the issue.
